docs/design/rubygems.md

```diff
@@ -0,0 +1,344 @@
+Using the bundler from the build image usually does not fit.
+
+## cachi2 implementation - TBD
+
+_The old way  of implementing new package managers to one big module is no longer prefered._
+_New package managers should split the logic into more self-contained modules wrapped in a package._
+
+### Vendoring solution
+
+Bundler has a built-in feature to cache all dependencies locally. This is done with the `bundle cache` command or `bundle package` alias.
+Default cache directory is `vendor/cache`.
+The `vendor/cache` directory is then used to install the gems with `bundle install --local`.
+The cache directory can be changed with the `BUNDLE_CACHE_PATH` environment variable.
+
+### Dependencies
+
+There four types of [sources](https://github.com/rubygems/rubygems/blob/master/bundler/lib/bundler/lockfile_parser.rb#L48) for dependencies in the `Gemfile.lock` file:
+
+#### Gem dependencies
+
+Regular gem dependencies located at source url, in our case always <https://rubygems.org>.
+Each gem can be accessed by its name and version - rubygems.org/gems/`<name>`-`<version>`.gem
+
+Example of a gem dependency in the `Gemfile.lock` file:
+
+```Gemfile.lock
+...
+GEM
+  remote: https://rubygems.org/
+  specs:
+    ...
+    rails (6.1.4)
+      # transitive dependencies
+      actioncable (= 6.1.4)
+      actionmailbox (= 6.1.4)
+      actionmailer (= 6.1.4)
+      actionpack (= 6.1.4)
+      actiontext (= 6.1.4)
+      actionview (= 6.1.4)
+      activejob (= 6.1.4)
+      activemodel (= 6.1.4)
+      activerecord (= 6.1.4)
+      activestorage (= 6.1.4)
+      activesupport (= 6.1.4)
+      bundler (>= 1.15.0)
+      railties (= 6.1.4)
+      sprockets-rails (>= 2.0.0)
+...
+```
+
+#### Git dependencies
+
+Example of a git dependency in the `Gemfile.lock` file:
+
+```Gemfile.lock
+...
+GIT
+  remote: https://github.com/porta.git
+  revision: 779beabd653afcd03c4468e0a69dc043f3bbb748
+  branch: main
+  specs:
+    porta (2.14.1)
+...
+```
+
+Bundler uses this [format](https://github.com/rubygems/rubygems/blob/3da9b1dda0824d1d770780352bb1d3f287cb2df5/bundler/lib/bundler/source/git.rb#L130) to cache git repositories:
+
+```ruby
+"#{base_name}-#{shortref_for_path(revision)}"
+```
+
+Any other format will cause bundler to re-download the repository -> cache invalidation -> the build will fail.
+
+#### Path dependencies
+
+Example of a path dependency in the `Gemfile.lock` file:
+
+```Gemfile.lock
+...
+PATH
+  remote: ../foobar
+  specs:
+    foobar (0.1.0)
+...
+```
+
+All path dependencies must be in the project directory.
+Bundler [does not copy](https://github.com/rubygems/rubygems/blob/master/bundler/lib/bundler/source/path.rb#L83)
+those dependencies that are already within the root directory of the project.
+
+#### Plugins
+
+Not supported by cachi2.
+
+### Checksums
+
+Checksum validation is enabled by default.
+It can be disabled with the `BUNDLE_DISABLE_CHECKSUM_VALIDATION` environment variable.
```

Comment:
So if I understand it correctly, bundler does do checksum validation even if there are no CHECKSUMS in the lockfile, but it just validates that the downloaded archive matches the checksum reported by the server (similar to how PyPI reports the expected checksums for archives and we use that to filter out non-matching wheels)

I realize now that we lose even _that_ small validation in Cachito, where we just don't validate any checksums at all. Which is not ideal, but we don't have to fix it right away in cachi2. A follow-up story to implement this checksum validation would be nice though.